golem-rib/src/parser/list_comprehension.rs

```diff
@@ -0,0 +1,103 @@
+// Copyright 2024 Golem Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::parser::errors::RibParseError;
+use crate::parser::identifier::identifier_text;
+use crate::parser::partial_block_expr::partial_block;
+use crate::parser::rib_expr::rib_expr as expr;
+use crate::{Expr, VariableId};
+use combine::parser::char::{alpha_num, char, spaces, string};
+use combine::{attempt, not_followed_by, optional, ParseError, Parser, Stream};
+
+pub fn list_comprehension<Input>() -> impl Parser<Input, Output = Expr>
+where
+    Input: Stream<Token = char>,
+    RibParseError: Into<
+        <Input::Error as ParseError<Input::Token, Input::Range, Input::Position>>::StreamError,
+    >,
+{
+    (
+        attempt(
+            string("for")
+                .skip(not_followed_by(alpha_num().or(char('-')).or(char('_'))).skip(spaces())),
```

Comment:
Such post logic can be thrown away once we fix after https://app.zenhub.com/workspaces/golem-cloud-65f5b68988aba80f7d690199/issues/gh/golemcloud/golem/971. It was super hard for me to avoid this with no tokenism

Reply:
The attempt is just in the beginning of token which I believe the path it need to backtrack is shorter and I can't see anything getting slower.